src/operator/linalg_impl.h

```diff
@@ -108,6 +112,55 @@ void linalg_gemm<gpu, DType>(const Tensor<gpu, 2, DType>& A, const Tensor<gpu, 2
 LINALG_GPU_GEMM(Sgemm, float)
 LINALG_GPU_GEMM(Dgemm, double)
 
+// Specialization of linalg_gemm<gpu, DType> for DType=mshadow::half::half_t.
+template<> inline
+void linalg_gemm<gpu, mshadow::half::half_t>(const Tensor<gpu, 2, mshadow::half::half_t>& A,
+                                             const Tensor<gpu, 2, mshadow::half::half_t>& B,
+                                             const Tensor<gpu, 2, mshadow::half::half_t>& C,
+                                             mshadow::half::half_t alpha,
+                                             mshadow::half::half_t beta,
+                                             bool tA, bool tB, Stream<gpu> *s) {
+  using namespace mxnet;
+  using mshadow::gpu;
+  CHECK_NOTNULL(s);
+  check_gemm(A, B, C, alpha, beta, tA, tB);
+
+  #if CUDA_VERSION >= 7050
+    auto blas_handle = Stream<gpu>::GetBlasHandle(s);
+    #if CUDA_VERSION >= 9000
+      auto cublas_math_mode = GetEnvAllowTensorCore() ? CUBLAS_TENSOR_OP_MATH
+                                                      : CUBLAS_DEFAULT_MATH;
+      auto previous_math_mode = SetCublasMathMode(blas_handle, cublas_math_mode);
```

Comment:
Sadly, a global setting performed at the creation of the blas handle has problems.  It would alter the behavior of Sgemm on fp32 I/O, introducing a cast-to-fp16 on the inputs to the multiplications.  The per-instance handle manipulations should be fast- the getenv() result is cached.